packages/framework/src/Model/Product/Product.php

```diff
@@ -723,9 +723,28 @@ public function markPriceAsRecalculated()
         $this->recalculatePrice = false;
     }
 
+    /**
+     * @param bool $recalculateVisibility
+     */
+    protected function setRecalculateVisibility($recalculateVisibility)
+    {
+        $this->recalculateVisibility = $recalculateVisibility;
+    }
+
     public function markForVisibilityRecalculation()
     {
-        $this->recalculateVisibility = true;
+        $this->setRecalculateVisibility(true);
+        if ($this->isMainVariant()) {
+            foreach ($this->getVariants() as $variant) {
+                $variant->setRecalculateVisibility(true);
+            }
+        } elseif ($this->isVariant()) {
+            $mainVariant = $this->getMainVariant();
+            // something must be read from $mainVariant, so Doctrine triggers lazy load
```

Comment:
Variant product have association to its main variant. This association is done by method `setMainVariant` or by Doctrine when the variant product is fetched from persistence. When the product is fetched from persistence, the `mainVariant` is only a proxy object, when we call something on this proxy object, Doctrine fetches it from persistence too.

And here we have a problem, the Doctrine seem to not fetch the main variant when we only write something into it, but when we read something first, Doctrine fetches the object, and the use-case works correctly.

If you think this is strange and it shouldn't work even before the code was moved to Product, you are right. But I tested that when the code is outside of Product, Doctrine does the job correctly, but once the code is inside of Product, Doctrine seem to not fetching the main variant. And one last blow - if the property `recalculateVisibility` is public, everything works correctly.

If you got it now and you agree with the solution, I will have to write better comment as you didn't get it earlier. If you disagree with the solution, please contact me, it is one crazy bug...

Reply:
ok, I get it now, it is crazy indeed :smile: To me, it is sufficient if you write the more descriptive comment. Maybe it would be nice to create an issue in Doctrine repository with detailed description and then link the issue in the comment :slightly_smiling_face:

Reply:
Well, Doctrine requires tests for issues, and it would take a lot of time isolating the problem. I think I shouldn't spent hours on creating an issue for Doctrine :-/

Reply:
I wrote maybe a better description

Reply:
:+1: